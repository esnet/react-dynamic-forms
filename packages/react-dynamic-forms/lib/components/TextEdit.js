"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _underscore = require("underscore");

var _underscore2 = _interopRequireDefault(_underscore);

var _revalidator = require("revalidator");

var _formGroup = require("../js/formGroup");

var _formGroup2 = _interopRequireDefault(_formGroup);

require("../css/textedit.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *  Copyright (c) 2015 - present, The Regents of the University of California,
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *  through Lawrence Berkeley National Laboratory (subject to receipt
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *  of any required approvals from the U.S. Dept. of Energy).
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *  All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *  This source code is licensed under the BSD-style license found in the
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *  LICENSE file in the root directory of this source tree.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

/**
 * Form control to edit a text field.
 * Set the initial value with `initialValue` and set a callback for
 * value changed with `onChange`.
 */
var TextEdit = function (_React$Component) {
    _inherits(TextEdit, _React$Component);

    function TextEdit(props) {
        _classCallCheck(this, TextEdit);

        var _this = _possibleConstructorReturn(this, (TextEdit.__proto__ || Object.getPrototypeOf(TextEdit)).call(this, props));

        _this.state = {
            touched: false
        };
        return _this;
    }

    _createClass(TextEdit, [{
        key: "isEmpty",
        value: function isEmpty(value) {
            return _underscore2.default.isNull(value) || _underscore2.default.isUndefined(value) || value === "";
        }
    }, {
        key: "isMissing",
        value: function isMissing(v) {
            return this.props.required && !this.props.disabled && this.isEmpty(v);
        }
    }, {
        key: "getError",
        value: function getError(value) {
            var result = { validationError: false, validationErrorMessage: null };

            // If the user has a field blank then that is never an error. Likewise if the field
            // is disabled then that is never an error.
            if (this.isEmpty(value) || this.props.disabled) {
                return result;
            }

            // Validate the value with Revalidator, given the rules in this.props.rules
            var obj = {};
            obj[this.props.name] = value;

            var properties = {};
            properties[this.props.name] = this.props.validation;

            var rules = this.props.validation ? { properties: properties } : null;
            if (obj && rules) {
                var validation = (0, _revalidator.validate)(obj, rules, { cast: true });
                var name = this.props.name || "Value";

                var msg = void 0;
                if (!validation.valid) {
                    msg = name + " " + validation.errors[0].message;
                    result.validationError = true;
                    result.validationErrorMessage = msg;
                }
            }
            return result;
        }
    }, {
        key: "componentWillReceiveProps",
        value: function componentWillReceiveProps(nextProps) {
            if (this.props.value !== nextProps.value) {
                var missing = this.isMissing(nextProps.value);

                var _getError = this.getError(nextProps.value),
                    validationError = _getError.validationError;

                // Broadcast error and missing states up to the owner


                if (this.props.onErrorCountChange) {
                    this.props.onErrorCountChange(this.props.name, validationError ? 1 : 0);
                }

                if (this.props.onMissingCountChange) {
                    this.props.onMissingCountChange(this.props.name, missing ? 1 : 0);
                }
            }
        }
    }, {
        key: "componentDidMount",
        value: function componentDidMount() {
            var missing = this.isMissing(this.props.value);

            var _getError2 = this.getError(this.props.value),
                validationError = _getError2.validationError;

            // Initial error and missing states are fed up to the owner


            if (this.props.onErrorCountChange) {
                this.props.onErrorCountChange(this.props.name, validationError ? 1 : 0);
            }

            if (this.props.onMissingCountChange) {
                this.props.onMissingCountChange(this.props.name, missing ? 1 : 0);
            }
        }
    }, {
        key: "handleChange",
        value: function handleChange() {
            var value = this.textInput.value;

            var missing = this.props.required && this.isEmpty(value);

            var _getError3 = this.getError(value),
                validationError = _getError3.validationError;

            var cast = value;

            // Callbacks
            if (this.props.onErrorCountChange) {
                this.props.onErrorCountChange(this.props.name, validationError ? 1 : 0);
            }

            if (this.props.onMissingCountChange) {
                this.props.onMissingCountChange(this.props.name, missing ? 1 : 0);
            }

            if (this.props.onChange) {
                if (_underscore2.default.has(this.props.rules, "type")) {
                    switch (this.props.rules.type) {
                        case "integer":
                            cast = value === "" ? null : parseInt(value, 10);
                            break;
                        case "number":
                            cast = value === "" ? null : parseFloat(value, 10);
                            break;
                        //pass
                        default:
                    }
                }
                this.props.onChange(this.props.name, cast);
            }
        }
    }, {
        key: "handleBlur",
        value: function handleBlur() {
            if (this.props.onBlur) {
                this.props.onBlur(this.props.name);
            }

            this.setState({ touched: true });
        }
    }, {
        key: "inlineStyle",
        value: function inlineStyle(hasError, isMissing) {
            var color = "";
            var background = "";
            if (hasError) {
                color = "#b94a48";
                background = "#fff0f3";
            } else if (isMissing) {
                background = "floralwhite";
            }
            return {
                color: color,
                background: background,
                // height: 23,
                height: "100%",
                width: "100%",
                paddingLeft: 3
            };
        }
    }, {
        key: "render",
        value: function render() {
            var _this2 = this;

            // Control state
            var isMissing = this.isMissing(this.props.value);

            var _getError4 = this.getError(this.props.value),
                validationError = _getError4.validationError,
                validationErrorMessage = _getError4.validationErrorMessage;

            if (this.props.edit) {
                // Error style/message
                var className = "";
                var msg = validationError && this.state.touched ? validationErrorMessage : "";
                var helpClassName = "help-block";
                if (validationError && this.state.touched) {
                    helpClassName += " has-error";
                    className = "has-error";
                }

                // Warning style
                var style = isMissing ? { background: "floralwhite" } : {};

                var type = this.props.type || "text";

                return _react2.default.createElement(
                    "div",
                    { className: className },
                    _react2.default.createElement("input", {
                        ref: function ref(input) {
                            _this2.textInput = input;
                        },
                        className: "form-control input-sm",
                        style: style,
                        type: type,
                        disabled: this.props.disabled,
                        placeholder: this.props.placeholder,
                        value: this.props.value,
                        onChange: function onChange() {
                            return _this2.handleChange();
                        },
                        onBlur: function onBlur() {
                            return _this2.handleBlur();
                        }
                    }),
                    _react2.default.createElement(
                        "div",
                        { className: helpClassName },
                        msg
                    )
                );
            } else {
                var view = this.props.view;
                var text = this.props.value;
                if (isMissing) {
                    text = " ";
                }
                var _style = this.inlineStyle(validationError, isMissing);
                if (!view) {
                    return _react2.default.createElement(
                        "div",
                        { style: _style },
                        text
                    );
                } else {
                    return _react2.default.createElement(
                        "div",
                        { style: _style },
                        view(text)
                    );
                }
            }
        }
    }]);

    return TextEdit;
}(_react2.default.Component);

exports.default = (0, _formGroup2.default)(TextEdit);