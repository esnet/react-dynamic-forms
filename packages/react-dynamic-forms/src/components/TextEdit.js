/**
 *  Copyright (c) 2015 - present, The Regents of the University of California,
 *  through Lawrence Berkeley National Laboratory (subject to receipt
 *  of any required approvals from the U.S. Dept. of Energy).
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree.
 */

import React from "react";
import _ from "underscore";
import { validate } from "revalidator";

import formGroup from "../js/formGroup";

import "../css/textedit.css";

/**
 * Form control to edit a text field.
 * Set the initial value with `initialValue` and set a callback for
 * value changed with `onChange`.
 */
class TextEdit extends React.Component {
    constructor(props) {
        super(props);
        this.state = { 
            touched: false 
        };
    }

    isEmpty(value) {
        return _.isNull(value) || _.isUndefined(value) || value === "";
    }

    isMissing(v) {
        return this.props.required && !this.props.disabled && this.isEmpty(v);
    }

    getError(value) {
        const result = { validationError: false, validationErrorMessage: null };

        // If the user has a field blank then that is never an error. Likewise if the field
        // is disabled then that is never an error.
        if (this.isEmpty(value) || this.props.disabled) {
            return result;
        }

        // Validate the value with Revalidator, given the rules in this.props.rules
        let obj = {};
        obj[this.props.name] = value;

        let properties = {};
        properties[this.props.name] = this.props.validation;

        const rules = this.props.validation ? { properties } : null;
        if (obj && rules) {
            const validation = validate(obj, rules, { cast: true });
            const name = this.props.name || "Value";

            let msg;
            if (!validation.valid) {
                msg = `${name} ${validation.errors[0].message}`;
                result.validationError = true;
                result.validationErrorMessage = msg;
            }
        }
        return result;
    }

    componentWillReceiveProps(nextProps) {
        if (this.props.value !== nextProps.value) {
            const missing = this.isMissing(nextProps.value);
            const { validationError } = this.getError(nextProps.value);

            // Broadcast error and missing states up to the owner
            if (this.props.onErrorCountChange) {
                this.props.onErrorCountChange(this.props.name, validationError ? 1 : 0);
            }

            if (this.props.onMissingCountChange) {
                this.props.onMissingCountChange(this.props.name, missing ? 1 : 0);
            }
        }
    }

    componentDidMount() {
        const missing = this.isMissing(this.props.value);
        const { validationError } = this.getError(this.props.value);

        // Initial error and missing states are fed up to the owner
        if (this.props.onErrorCountChange) {
            this.props.onErrorCountChange(this.props.name, validationError ? 1 : 0);
        }

        if (this.props.onMissingCountChange) {
            this.props.onMissingCountChange(this.props.name, missing ? 1 : 0);
        }
    }

    handleChange() {
        const { value } = this.textInput;
        const missing = this.props.required && this.isEmpty(value);
        const { validationError } = this.getError(value);
        let cast = value;

        // Callbacks
        if (this.props.onErrorCountChange) {
            this.props.onErrorCountChange(this.props.name, validationError ? 1 : 0);
        }

        if (this.props.onMissingCountChange) {
            this.props.onMissingCountChange(this.props.name, missing ? 1 : 0);
        }

        if (this.props.onChange) {
            if (_.has(this.props.rules, "type")) {
                switch (this.props.rules.type) {
                    case "integer":
                        cast = value === "" ? null : parseInt(value, 10);
                        break;
                    case "number":
                        cast = value === "" ? null : parseFloat(value, 10);
                        break;
                    //pass
                    default:
                }
            }
            this.props.onChange(this.props.name, cast);
        }
    }

    handleBlur() {
        if (this.props.onBlur) {
            this.props.onBlur(this.props.name);
        }

        this.setState({ touched: true });
    }

    inlineStyle(hasError, isMissing) {
        let color = "";
        let background = "";
        if (hasError) {
            color = "#b94a48";
            background = "#fff0f3";
        } else if (isMissing) {
            background = "floralwhite";
        }
        return {
            color,
            background,
            // height: 23,
            height: "100%",
            width: "100%",
            paddingLeft: 3
        };
    }

    render() {
        // Control state
        const isMissing = this.isMissing(this.props.value);
        const { validationError, validationErrorMessage } = this.getError(this.props.value);

        if (this.props.edit) {
            // Error style/message
            let className = "";
            const msg = validationError && this.state.touched ? validationErrorMessage : "";
            let helpClassName = "help-block";
            if (validationError && this.state.touched) {
                helpClassName += " has-error";
                className = "has-error";
            }

            // Warning style
            const style = isMissing ? { background: "floralwhite" } : {};

            const type = this.props.type || "text";

            return (
                <div className={className}>
                    <input
                        ref={(input) => { this.textInput = input; }}
                        className="form-control input-sm"
                        style={style}
                        type={type}
                        disabled={this.props.disabled}
                        placeholder={this.props.placeholder}
                        value={this.props.value}
                        onChange={() => this.handleChange()}
                        onBlur={() => this.handleBlur()}
                    />
                    <div className={helpClassName}>{msg}</div>
                </div>
            );
        } else {
            const view = this.props.view;
            let text = this.props.value;
            if (isMissing) {
                text = " ";
            }
            const style = this.inlineStyle(validationError, isMissing);
            if (!view) {
                return <div style={style}>{text}</div>;
            } else {
                return <div style={style}>{view(text)}</div>;
            }
        }
    }
}

export default formGroup(TextEdit);
