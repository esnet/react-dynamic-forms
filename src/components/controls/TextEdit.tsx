/**
 *  Copyright (c) 2015 - present, The Regents of the University of California,
 *  through Lawrence Berkeley National Laboratory (subject to receipt
 *  of any required approvals from the U.S. Dept. of Energy).
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree.
 */

import Flexbox from "@g07cha/flexbox-react";
import _ from "lodash";
import React from "react";
import Form from "react-bootstrap/Form";
import { validate } from "revalidator";
import { formGroup } from "../../hoc/group";
import "../../style/textedit.css";
import { editAction } from "../../util/actions";
import { textView } from "../../util/renderers";
import {
    colors,
    inlineCancelButtonStyle,
    inlineDoneButtonStyle,
    inlineStyle
} from "../../util/style";
import { FieldEditorProps, FieldValue } from "../Form";

interface ValidationError {
    validationError: boolean;
    validationErrorMessage: string | null;
}

interface Props {
    field: string;
    type: string;
    rules: any;
    view: (value: FieldValue) => React.ReactElement<any>;
}

export interface TextEditState {
    value: FieldValue;
    oldValue: FieldValue;
    isFocused: boolean;
    touched: boolean;
    selectText: boolean;
    hover: boolean;
}

export type TextEditProps = Props & FieldEditorProps;

/**
 * Form control to edit a text field.
 */
class TextEdit extends React.Component<TextEditProps, TextEditState> {
    textInput: any;
    constructor(props: TextEditProps) {
        super(props);
        this.state = {
            value: props.value,
            oldValue: null,
            isFocused: false,
            touched: false,
            selectText: false,
            hover: false
        };

        this.handleChange = this.handleChange.bind(this);
        this.handleKeyPress = this.handleKeyPress.bind(this);
        this.handleFocus = this.handleFocus.bind(this);
    }

    handleMouseEnter() {
        this.setState({ hover: true });
    }

    handleMouseLeave() {
        this.setState({ hover: false });
    }

    handleEditItem() {
        if (this.props.onEditItem) {
            this.props.onEditItem(this.props.name);
        }
    }

    isEmpty(value: FieldValue) {
        return _.isNull(value) || _.isUndefined(value) || value === "";
    }

    isMissing(value: FieldValue) {
        return this.props.required && !this.props.disabled && this.isEmpty(value);
    }

    getError(value: FieldValue) {
        const result: ValidationError = {
            validationError: false,
            validationErrorMessage: null
        };

        // If the user has a field blank then that is never an error. Likewise if the field
        // is disabled then that is never an error.
        if (this.isEmpty(value) || this.props.disabled) {
            return result;
        }

        // Validate the value with Revalidator, given the rules in this.props.rules
        let obj = {};
        obj[this.props.name] = value;

        let properties = {};
        properties[this.props.name] = this.props.validation;

        const rules = this.props.validation ? { properties } : null;
        if (obj && rules) {
            const validation = validate(obj, rules, { cast: true });
            const name = this.props.name || "Value";

            let msg;
            if (!validation.valid) {
                msg = `${name} ${validation.errors[0].message}`;
                result.validationError = true;
                result.validationErrorMessage = msg;
            }
        }
        return result;
    }

    componentWillReceiveProps(nextProps: TextEditProps) {
        if (nextProps.selected && this.props.edit !== nextProps.edit && nextProps.edit === true) {
            this.setState({ selectText: true });
        }
        if (this.state.value !== nextProps.value && !this.state.isFocused) {
            this.setState({ value: nextProps.value });

            const missing = this.isMissing(nextProps.value);
            const { validationError: error } = this.getError(nextProps.value);

            // Broadcast error and missing states up to the parent
            if (this.props.onErrorCountChange) {
                this.props.onErrorCountChange(this.props.name, error ? 1 : 0);
            }
            if (this.props.onMissingCountChange) {
                this.props.onMissingCountChange(this.props.name, missing ? 1 : 0);
            }
        }
    }

    componentDidMount() {
        const missing = this.isMissing(this.props.value);
        const { validationError } = this.getError(this.props.value);

        // Initial error and missing states are fed up to the parent
        if (this.props.onErrorCountChange) {
            this.props.onErrorCountChange(this.props.name, validationError ? 1 : 0);
        }

        if (this.props.onMissingCountChange) {
            this.props.onMissingCountChange(this.props.name, missing ? 1 : 0);
        }
    }

    componentDidUpdate() {
        if (this.state.selectText) {
            this.textInput.focus();
            this.textInput.select();
            this.setState({ selectText: false });
        }
    }

    handleChange(e: React.FormEvent<any>): void {
        const name = this.props.name;
        const value: string = (e.target as HTMLInputElement).value;

        this.setState({ value }, () => {
            const missing = this.props.required && this.isEmpty(value);
            const { validationError } = this.getError(value);
            let cast: FieldValue = value;

            // Callbacks
            if (this.props.onErrorCountChange) {
                this.props.onErrorCountChange(name, validationError ? 1 : 0);
            }

            if (this.props.onMissingCountChange) {
                this.props.onMissingCountChange(name, missing ? 1 : 0);
            }

            if (this.props.onChange) {
                if (_.has(this.props.rules, "type")) {
                    switch (this.props.rules.type) {
                        case "integer":
                            cast = value === "" ? null : parseInt(value, 10);
                            break;
                        case "number":
                            cast = value === "" ? null : parseFloat(value);
                            break;
                        //pass
                        default:
                    }
                }
                this.props.onChange(name, cast);
            }
        });
    }

    handleFocus(): void {
        if (!this.state.isFocused) {
            this.setState({ isFocused: true, oldValue: this.props.value });
        }
    }

    handleKeyPress(e: React.KeyboardEvent<HTMLInputElement>): void {
        if (e.key === "Enter") {
            if (!e.shiftKey) {
                this.handleDone();
            }
        }
        if (e.keyCode === 27 /* ESC */) {
            this.handleCancel();
        }
    }

    handleDone() {
        if (this.props.onBlur) {
            this.props.onBlur(this.props.name);
        }
        this.setState({ isFocused: false, hover: false, oldValue: null });
    }

    handleCancel() {
        if (this.props.onChange) {
            const v = this.state.oldValue;
            let cast: FieldValue = v;
            if (_.has(this.props.rules, "type")) {
                // If we have a rule for the value to be either an interger or a number
                // then we parse the string to either an Int or Float and set the onChange
                // value to that
                if (_.isString(v)) {
                    switch (this.props.rules.type) {
                        case "integer":
                            cast = v === "" ? null : parseInt(v, 10);
                            break;
                        case "number":
                            cast = v === "" ? null : parseFloat(v);
                            break;
                        //pass
                        default:
                    }
                }
            }
            this.props.onChange(this.props.name, cast);
        }
        this.props.onBlur(this.props.name);
        this.setState({ isFocused: false, hover: false, oldValue: null });
    }

    render() {
        // Control state
        const isMissing = this.isMissing(this.props.value);
        const { validationError, validationErrorMessage } = this.getError(this.props.value);

        if (this.props.edit) {
            // Error style/message
            let className = "";
            const msg = validationError && this.state.touched ? validationErrorMessage : "";
            let helpClassName = "help-block";
            if (validationError && this.state.touched) {
                helpClassName += " has-error";
                className = "has-error";
            }

            // Warning style
            const style = isMissing ? { background: colors.MISSING_COLOR_BG } : {};
            const type = this.props.type || "text";

            const canCommit = !isMissing && !validationError;

            return (
                <Flexbox flexDirection="row" style={{ width: "100%" }}>
                    <div className={className} style={{ width: "100%" }}>
                        <Form.Control
                            value={this.state.value ? `${this.state.value}` : ""}
                            size="sm"
                            ref={(input: any) => {
                                this.textInput = input;
                            }}
                            style={style}
                            type={type}
                            disabled={this.props.disabled}
                            placeholder={this.props.placeholder}
                            onChange={this.handleChange}
                            onFocus={this.handleFocus}
                            onKeyUp={this.handleKeyPress}
                        />
                        <div className={helpClassName}>{msg}</div>
                    </div>
                    {this.props.selected ? (
                        <span style={{ marginTop: 3 }}>
                            {canCommit ? (
                                <span
                                    style={inlineDoneButtonStyle(5, true)}
                                    onClick={() => this.handleDone()}
                                >
                                    DONE
                                </span>
                            ) : (
                                <span style={inlineDoneButtonStyle(5, false)}>DONE</span>
                            )}

                            <span
                                style={inlineCancelButtonStyle()}
                                onClick={() => this.handleCancel()}
                            >
                                CANCEL
                            </span>
                        </span>
                    ) : (
                        <div />
                    )}
                </Flexbox>
            );
        } else {
            const view = this.props.view || textView;
            const text = isMissing ? (
                <span />
            ) : (
                <span style={{ minHeight: 28 }}>{view(this.props.value)}</span>
            );
            const edit = editAction(this.state.hover && this.props.allowEdit, () =>
                this.handleEditItem()
            );
            const style = inlineStyle(validationError, isMissing);
            return (
                <div
                    style={style}
                    key={`key-${isMissing}-${validationError}`}
                    onMouseEnter={() => this.handleMouseEnter()}
                    onMouseLeave={() => this.handleMouseLeave()}
                >
                    {text}
                    {edit}
                </div>
            );
        }
    }
}

export default formGroup(TextEdit);
