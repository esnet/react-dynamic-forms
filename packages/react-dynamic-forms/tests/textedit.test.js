/**
 *  Copyright (c) 2017 - present, The Regents of the University of California,
 *  through Lawrence Berkeley National Laboratory (subject to receipt
 *  of any required approvals from the U.S. Dept. of Energy).
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree.
 */

import renderer from "react-test-renderer";
import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

import {
    TestBasic,
    TextEditDisabled,
    TextEditPlaceholder,
    TextEditRequired1,
    TextEditRequired2,
    TextEditRequired3,
    TextEditValidate,
    TextEditValidateInt
} from "./components_test/TextEditExamples";

import TextEdit from "../src/components/TextEdit";

//
// Visual appearance snapshot tests
//
test("TextEdit with an initial value and width", () => {
    const tree = renderer.create(<TestBasic />).toJSON();
    expect(tree).toMatchSnapshot();
});

test("TextEdit that is disabled", () => {
    const tree = renderer.create(<TextEditDisabled />).toJSON();
    expect(tree).toMatchSnapshot();
});

test("TextEdit with a placeholder", () => {
    const tree = renderer.create(<TextEditPlaceholder />).toJSON();
    expect(tree).toMatchSnapshot();
});

test("TextEdit with required field (with showRequired turned ON)", () => {
    const tree = renderer.create(<TextEditRequired1 />).toJSON();
    expect(tree).toMatchSnapshot();
});

test("TextEdit with required field (with showRequired turned OFF)", () => {
    const tree = renderer.create(<TextEditRequired2 />).toJSON();
    expect(tree).toMatchSnapshot();
});

test("Required field (with showRequired turned ON and initial value", () => {
    const tree = renderer.create(<TextEditRequired3 />).toJSON();
    expect(tree).toMatchSnapshot();
});

test("Validated field (email address)", () => {
    const tree = renderer.create(<TextEditValidate />).toJSON();
    expect(tree).toMatchSnapshot();
});

test("Validated field (integer)", () => {
    const tree = renderer.create(<TextEditValidateInt />).toJSON();
    expect(tree).toMatchSnapshot();
});

//
// Test callbacks
//
test("TextEdit onChange callback", () => {
    const handleChange = jest.fn();

    const wrapper = mount(
        <TextEdit attr="mytextedit" initialValue="bob" onChange={handleChange} />
    );

    // Simulate user typing
    const input = wrapper.find("input");
    input.node.value = "fred";
    input.simulate("blur");

    // Get the result of the handleChange callback
    const [attr, value] = handleChange.mock.calls[0];
    expect(handleChange).toHaveBeenCalled();
    expect(attr).toBe("mytextedit");
    expect(value).toBe("fred");
});

test("TextEdit isRequired callback", () => {
    const handleMissingCountChange = jest.fn();

    const wrapper = mount(
        <TextEdit
            attr="mytextedit"
            initialValue="bob"
            required={true}
            onMissingCountChange={handleMissingCountChange}
        />
    );

    // The TextEdit will evoke onMissingCountChange initially to report
    // up initial missing value counts, so handleMissingCountChange
    // should have been called once at this point, and since we provided
    // an initialValue the missing count should be 0
    const [attr, missingBefore] = handleMissingCountChange.mock.calls[0];
    expect(handleMissingCountChange.mock.calls.length).toBe(1);
    expect(attr).toBe("mytextedit");
    expect(missingBefore).toBe(0);

    // Simulate the user clearing the input field
    const input = wrapper.find("input");
    input.node.value = "";
    input.simulate("blur");

    // Get the result of the handleChange callback
    expect(handleMissingCountChange.mock.calls.length).toBe(2);
    const [_, missingAfter] = handleMissingCountChange.mock.calls[1];
    expect(missingAfter).toBe(1);
});

test("TextEdit isRequired snapshot test", () => {
    const handleMissingCountChange = jest.fn();

    const wrapper = mount(
        <TextEdit
            attr="mytextedit"
            initialValue="bob"
            required={true}
            showRequired={true}
            onMissingCountChange={handleMissingCountChange}
        />
    );
    const input = wrapper.find("input");
    const div = wrapper.find("div");

    // Get an initial snapshot of the TextEdit
    expect(toJson(div)).toMatchSnapshot();

    // Check if it shows an error (it shouldn't yet)
    expect(div.node.className).toBe("");

    // Simulate the user clearing the input field
    input.node.value = "";
    input.simulate("blur");

    // Check if it shows an error now
    expect(div.node.className).toBe("has-error");

    // Get an after snapshot of the TextEdit
    expect(toJson(div)).toMatchSnapshot();
});

test("TextEdit email address validate", () => {
    const handleErrorCountChange = jest.fn();

    const wrapper = mount(
        <TextEdit
            initialValue="bob.at.gmail.com"
            onErrorCountChange={handleErrorCountChange}
            rules={{ format: "email" }}
        />
    );
    const input = wrapper.find("input");
    const div = wrapper.find("div");

    const helptext = div.find(".has-error");
    expect(helptext.node.textContent).toEqual("Value is not a valid email");

    // Get an initial snapshot of the TextEdit
    expect(toJson(div)).toMatchSnapshot();

    // Check if it shows an error (it should because it starts with an invalid email)
    expect(div.node.className).toBe("has-error");

    // Get the result of the initial call to the error count callback
    expect(handleErrorCountChange.mock.calls.length).toBe(1);
    const [, errorBefore] = handleErrorCountChange.mock.calls[0];
    expect(errorBefore).toBe(1);

    // Simulate the user clearing the input field
    input.node.value = "bob@gmail.com";
    input.simulate("blur");

    // Check if it shows an error now
    expect(div.node.className).toBe("");

    // Get an after snapshot of the TextEdit
    expect(toJson(div)).toMatchSnapshot();

    // Get the result of the error change callback
    expect(handleErrorCountChange.mock.calls.length).toBe(2);
    const [, errorAfter] = handleErrorCountChange.mock.calls[1];
    expect(errorAfter).toBe(0);
});
