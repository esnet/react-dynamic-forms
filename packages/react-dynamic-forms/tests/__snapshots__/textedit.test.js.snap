exports[`test Required field (with showRequired turned ON and initial value 1`] = `
<div
  className="">
  <input
    className="form-control input-sm"
    defaultValue="Bob"
    disabled={undefined}
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text" />
  <div
    className="help-block">
    
  </div>
</div>
`;

exports[`test TextEdit email address validate 1`] = `
<div
  className="has-error">
  <input
    className="form-control input-sm"
    defaultValue="bob.at.gmail.com"
    onBlur={[Function]}
    onFocus={[Function]}
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text" />
  <div
    className="help-block has-error">
    Value is not a valid email
  </div>
</div>
`;

exports[`test TextEdit email address validate 2`] = `
<div
  className="">
  <input
    className="form-control input-sm"
    defaultValue="bob@gmail.com"
    onBlur={[Function]}
    onFocus={[Function]}
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text" />
  <div
    className="help-block" />
</div>
`;

exports[`test TextEdit isRequired snapshot test 1`] = `
<div
  className="">
  <input
    className="form-control input-sm"
    defaultValue="bob"
    onBlur={[Function]}
    onFocus={[Function]}
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text" />
  <div
    className="help-block" />
</div>
`;

exports[`test TextEdit isRequired snapshot test 2`] = `
<div
  className="has-error">
  <input
    className="form-control input-sm"
    defaultValue=""
    onBlur={[Function]}
    onFocus={[Function]}
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text" />
  <div
    className="help-block" />
</div>
`;

exports[`test TextEdit that is disabled 1`] = `
<div
  className="">
  <input
    className="form-control input-sm"
    defaultValue="Bob"
    disabled={true}
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    required={true}
    style={
      Object {
        "width": 300,
      }
    }
    type="text" />
  <div
    className="help-block">
    
  </div>
</div>
`;

exports[`test TextEdit with a placeholder 1`] = `
<div
  className="">
  <input
    className="form-control input-sm"
    defaultValue={undefined}
    disabled={undefined}
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder="Enter first name"
    required={true}
    style={
      Object {
        "width": 300,
      }
    }
    type="text" />
  <div
    className="help-block">
    
  </div>
</div>
`;

exports[`test TextEdit with an initial value and width 1`] = `
<div
  className="">
  <input
    className="form-control input-sm"
    defaultValue="Bob"
    disabled={undefined}
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    required={true}
    style={
      Object {
        "width": 300,
      }
    }
    type="text" />
  <div
    className="help-block">
    
  </div>
</div>
`;

exports[`test TextEdit with required field (with showRequired turned OFF) 1`] = `
<div
  className="">
  <input
    className="form-control input-sm"
    defaultValue={undefined}
    disabled={undefined}
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text" />
  <div
    className="help-block">
    
  </div>
</div>
`;

exports[`test TextEdit with required field (with showRequired turned ON) 1`] = `
<div
  className="has-error">
  <input
    className="form-control input-sm"
    defaultValue={undefined}
    disabled={undefined}
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text" />
  <div
    className="help-block">
    
  </div>
</div>
`;

exports[`test Validated field (email address) 1`] = `
<div
  className="has-error">
  <input
    className="form-control input-sm"
    defaultValue="bob.at.gmail.com"
    disabled={undefined}
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text" />
  <div
    className="help-block has-error">
    Value is not a valid email
  </div>
</div>
`;

exports[`test Validated field (integer) 1`] = `
<div
  className="">
  <input
    className="form-control input-sm"
    defaultValue="42"
    disabled={undefined}
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text" />
  <div
    className="help-block">
    
  </div>
</div>
`;
